model.py

```diff
@@ -218,6 +226,190 @@ def nearby(cls, _db, latitude, longitude, max_radius=150):
                 distance.asc())
         return qu
 
+    @classmethod
+    def search(cls, _db, latitude, longitude, query):
+        """Try as hard as possible to find a small number of libraries
+        that match the given query.
+
+        Preference will be given to libraries closer to the current
+        latitude/longitude.
+        """
+        # We don't anticipate a lot of libraries or a lot of
+        # localities with the same name, but we need to have _some_
+        # kind of limit just to place an upper bound on how bad things
+        # can get. This will guarantee we never return more than 20
+        # results.
+        max_libraries = 10
+        
+        if not query:
+            # No query, no results.
+            return []
+        here = GeometryUtility.point(latitude, longitude)
+
+        library_query, place_query, place_type = cls.query_parts(query)
+
+        # We start with libraries that match the name query.
+        if library_query:
+            libraries_for_name = cls.search_by_library_name(
+                _db, library_query, here).limit(max_libraries).all()
+        else:
+            libraries_for_name = []
+            
+        # We tack on any additional libraries that match a place query.
+        if place_query:
+            libraries_for_location = cls.search_by_location_name(
+                _db, place_query, place_type, here,
+                libraries_for_name
+            ).limit(max_libraries).all()
+        else:
+            libraries_for_location = []
+            
+        return libraries_for_name + libraries_for_location
```

Comment:
If a library matches both queries, it should only appear once in the final list.